src/dotnet/PrintableTable.cs

```diff
@@ -0,0 +1,199 @@
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using Microsoft.DotNet.Tools;
+
+namespace Microsoft.DotNet.Cli
+{
+    // Represents a table (with rows of type T) that can be printed to a terminal.
+    internal class PrintableTable<T>
+    {
+        private const string ColumnDelimiter = "      ";
+        private List<Column> _columns = new List<Column>();
+
+        private class Column
+        {
+            public string Header { get; set; }
+            public Func<T, string> GetContent { get; set; }
+            public int MaxWidth { get; set; }
+            public override string ToString() { return Header; }
+        }
+
+        public void AddColumn(string header, Func<T, string> getContent, int maxWidth = int.MaxValue)
+        {
+            if (string.IsNullOrEmpty(header))
+            {
+                throw new ArgumentException(
+                    CommonLocalizableStrings.ArgumentNullOrEmpty,
+                    nameof(header));
+            }
+
+            if (getContent == null)
+            {
+                throw new ArgumentNullException(nameof(getContent));
+            }
+
+            if (maxWidth <= 0)
+            {
+                throw new ArgumentException(
+                    CommonLocalizableStrings.ColumnMaxWidthMustBeGreaterThanZero,
+                    nameof(maxWidth));
+            }
+
+            _columns.Add(
+                new Column() {
+                    Header = header,
+                    GetContent = getContent,
+                    MaxWidth = maxWidth
+                });
+        }
+
+        public void Print(IEnumerable<T> rows, Action<string> writeLine)
+        {
+            if (rows == null)
+            {
+                throw new ArgumentNullException(nameof(rows));
+            }
+
+            if (writeLine == null)
+            {
+                throw new ArgumentNullException(nameof(writeLine));
+            }
+
+            if (_columns.Count == 0)
+            {
+                return;
+            }
+
+            var widths = CalculateColumnWidths(rows);
+
+            foreach (var line in EnumerateHeaderLines(widths))
+            {
+                writeLine(line);
+            }
+
+            writeLine(new string('-', GetTotalWidth(widths)));
+
+            foreach (var row in rows)
+            {
+                foreach (var line in EnumerateRowLines(row, widths))
+                {
+                    writeLine(line);
+                }
+            }
+        }
+
+        public int CalculateWidth(IEnumerable<T> rows)
+        {
+            if (rows == null)
+            {
+                throw new ArgumentNullException(nameof(rows));
+            }
+
+            return GetTotalWidth(CalculateColumnWidths(rows));
+        } 
+
+        private IEnumerable<string> EnumerateHeaderLines(int[] widths)
+        {
```

Comment:
Quick look at the code appears that only one header line is supported, so is the name and IEnumerable correct?

Reply:
Depending on the column width, the column headers may also wrap across multiple lines.